### ISSUE TITLE
pgr_bdDijkstra with non-existing vid gives segmentation fault and sets database system in recovery mode

### ISSUE THREAD

**Issue by RinkeHeida** [pgRouting/pgrouting#2316]
pgr_bdDijkstra(edges_sql, start_vid,  end_vid  [, directed]) results in a segmentation fault and setting database in recovery mode when start_vid and/or end_vid do not exist.

**Version info**

select version()
` PostgreSQL 13.3 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit`
 
select postgis_full_version()
` POSTGIS="3.0.3 6660953" [EXTENSION] PGSQL="130" GEOS="3.8.0-CAPI-1.13.1 " PROJ="Rel. 5.2.0, September 15th, 2018" LIBXML="2.9.1" LIBJSON="0.13.1" LIBPROTOBUF="1.3.2" WAGYU="0.4.3 (Internal)"
` 
select pgr_version()
` 3.1.0`

**Create sample data**

```
CREATE TABLE edge_table (
    id BIGSERIAL,
    dir character varying,
    source BIGINT,
    target BIGINT,
    cost FLOAT,
    reverse_cost FLOAT,
    capacity BIGINT,
    reverse_capacity BIGINT,
    category_id INTEGER,
    reverse_category_id INTEGER,
    x1 FLOAT,
    y1 FLOAT,
    x2 FLOAT,
    y2 FLOAT,
    the_geom geometry
);
```

```
INSERT INTO edge_table (
    category_id, reverse_category_id,
    cost, reverse_cost,
    capacity, reverse_capacity,
    x1, y1,
    x2, y2) VALUES
(3, 1,    1,  1,  80, 130,   2,   0,    2, 1),
(3, 2,   -1,  1,  -1, 100,   2,   1,    3, 1),
(2, 1,   -1,  1,  -1, 130,   3,   1,    4, 1),
(2, 4,    1,  1, 100,  50,   2,   1,    2, 2),
(1, 4,    1, -1, 130,  -1,   3,   1,    3, 2),
(4, 2,    1,  1,  50, 100,   0,   2,    1, 2),
(4, 1,    1,  1,  50, 130,   1,   2,    2, 2),
(2, 1,    1,  1, 100, 130,   2,   2,    3, 2),
(1, 3,    1,  1, 130,  80,   3,   2,    4, 2),
(1, 4,    1,  1, 130,  50,   2,   2,    2, 3),
(1, 2,    1, -1, 130,  -1,   3,   2,    3, 3),
(2, 3,    1, -1, 100,  -1,   2,   3,    3, 3),
(2, 4,    1, -1, 100,  -1,   3,   3,    4, 3),
(3, 1,    1,  1,  80, 130,   2,   3,    2, 4),
(3, 4,    1,  1,  80,  50,   4,   2,    4, 3),
(3, 3,    1,  1,  80,  80,   4,   1,    4, 2),
(1, 2,    1,  1, 130, 100,   0.5, 3.5,  1.999999999999,3.5),
(4, 1,    1,  1,  50, 130,   3.5, 2.3,  3.5,4);
```

```
UPDATE edge_table SET the_geom = st_makeline(st_point(x1,y1),st_point(x2,y2)),
dir = CASE WHEN (cost>0 AND reverse_cost>0) THEN 'B'   -- both ways
           WHEN (cost>0 AND reverse_cost<0) THEN 'FT'  -- direction of the LINESSTRING
           WHEN (cost<0 AND reverse_cost>0) THEN 'TF'  -- reverse direction of the LINESTRING
           ELSE '' END;                                -- unknown

SELECT pgr_createTopology('edge_table',0.001);
```

**Correct results**
```
SELECT * FROM pgr_bdDijkstra(
    'SELECT id, source, target, cost, reverse_cost FROM edge_table',
    2, 3
);
```
Gives same result table as on the pgrouting website

**Generate fault**
Note: _do this in a very small database, otherwise PostgreSQL can be hours in recovery mode, not accepting any connections_

Use non-exising start_vid and/or end_vid:

```
SELECT * FROM pgr_bdDijkstra(
    'SELECT id, source, target, cost, reverse_cost FROM edge_table',
    2, 20
);
```

Logfile fragments:
```
2022-06-15 14:06:03 UTC::@:[7855]:LOG: server process (PID 5406) was terminated by signal 11: Segmentation fault
2022-06-15 14:06:03 UTC::@:[7855]:DETAIL: Failed process was running: SELECT * FROM pgr_bdDijkstra(
'SELECT id, source, target, cost, reverse_cost FROM edge_table',
2, 20
)2022-06-15 14:06:03 UTC::@:[7855]:LOG: terminating any other active server processes
2022-06-15 14:06:03 UTC:10.98.85.65(50686):<user>@postgres:[5448]:WARNING: terminating connection because of crash of another server process
2022-06-15 14:06:03 UTC:10.98.85.65(50686):<user>@postgres:[5448]:DETAIL: The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2022-06-15 14:06:03 UTC:10.98.85.65(50686):<user>@postgres:[5448]:HINT: In a moment you should be able to reconnect to the database and repeat your command.
2022-06-15 14:06:03 UTC:10.98.85.65(50684):<user>@<database>:[5440]:WARNING: terminating connection because of crash of another server process
2022-06-15 14:06:03 UTC:10.98.85.65(50684):<user>@<database>:[5440]:DETAIL: The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2022-06-15 14:06:03 UTC:10.98.85.65(50684):<user>@<database>:[5440]:HINT: In a moment you should be able to reconnect to the database and repeat your command.
2022-06-15 14:06:03 UTC::@:[5305]:WARNING: terminating connection because of crash of another server process
2022-06-15 14:06:03 UTC::@:[5305]:DETAIL: The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2022-06-15 14:06:03 UTC::@:[5305]:HINT: In a moment you should be able to reconnect to the database and repeat your command.

2022-06-15 14:06:03 UTC:10.98.85.65(50680):<user>@<database>:[5359]:DETAIL: The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2022-06-15 14:06:03 UTC:10.98.85.65(50680):<user>@<database>:[5359]:HINT: In a moment you should be able to reconnect to the database and repeat your command.
2022-06-15 14:06:03 UTC:10.98.85.65(50685):<user>@postgres:[5444]:WARNING: terminating connection because of crash of another server process
2022-06-15 14:06:03 UTC:10.98.85.65(50685):<user>@postgres:[5444]:DETAIL: The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2022-06-15 14:06:03 UTC:10.98.85.65(50685):<user>@postgres:[5444]:HINT: In a moment you should be able to reconnect to the database and repeat your command.
2022-06-15 14:06:03 UTC::@:[7855]:LOG: all server processes terminated; reinitializing
2022-06-15 14:06:03 UTC::@:[5452]:LOG: database system was interrupted; last known up at 2022-06-15 14:05:57 UTC
2022-06-15 14:06:03 UTC::@:[5452]:LOG: database system was not properly shut down; automatic recovery in progress
2022-06-15 14:06:03 UTC::@:[5452]:LOG: invalid record length at 12AE/F40027D0: wanted 24, got 0
2022-06-15 14:06:03 UTC::@:[5452]:LOG: redo is not required
2022-06-15 14:06:03 UTC:10.98.85.65(50687):<user>@<database>:[5454]:FATAL: the database system is in recovery mode
2022-06-15 14:06:03 UTC:10.98.85.65(50688):<user>@<database>:[5455]:FATAL: the database system is in recovery mode
2022-06-15 14:06:03 UTC::@:[5452]:LOG: checkpoint starting: end-of-recovery immediate
2022-06-15 14:06:03 UTC::@:[5452]:LOG: checkpoint complete: wrote 0 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.007 s, sync=0.001 s, total=0.015 s; sync files=0, longest=0.000 s, average=0.000 s; distance=0 kB, estimate=0 kB
2022-06-15 14:06:03 UTC::@:[7855]:LOG: database system is ready to accept connections
```
**Same query with pg_Dijkstra**
Replacing pgr_bdDijkstra with pgr_Dijkstra:

```
SELECT * FROM pgr_Dijkstra(
    'SELECT id, source, target, cost, reverse_cost FROM edge_table',
    2, 20
);
```
Gives empty result set (as expected)
So this problem specifically exists in pgr_bdDijkstra.

**Comment by cvvergara** [comment]
Can you proved the results of:
```
SELECT * FROM pgr_full_version();
```

**Comment by RinkeHeida** [comment]
Hi Vicky,

SELECT * FROM pgr_full_version();

|version|build_type|compile_date|library |system |postgresql |compiler 
|boost |hash |
|-------|----------|------------|---------------|--------------------------------|---------------|---------|------|-------|
|3.1.0 |Release |2021/06/14 
|pgrouting-3.1.0|Linux-4.14.59-64.43.amzn1.x86_64|PostgreSQL 
13.3|GNU-4.8.5|1.65.0|unknown|

Regards,

Rinke

On 2022-06-17 20:51, Vicky Vergara wrote:

> Can you proved the results of:
> 
> SELECT * FROM pgr_full_version();
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe 
> [2].
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>



Links:
------
[1] 
https://github.com/pgRouting/pgrouting/issues/2316#issuecomment-1159148093
[2]

**Comment by cvvergara** [comment]
I am trying to reproduce your problem 
But my library version is `pgrouting-3.1.3` and yours is `pgrouting-3.1.0`
Aka the SQL code I am using is the same you are using, But the C++ code I am using is from `pgrouting-3.1.3` and yours is from `pgrouting-3.1.0`

Also the boost version is `1.65.0` and the one that I tested on is `1.71.0`

So with what I am using I don't get the problem:
```
sampledata=# select * FROM pgr_full_version();                                              
 version | build_type | compile_date |     library     |         system          |                 postgresql                  | compiler  | boost  |    hash    
---------+------------+--------------+-----------------+-------------------------+---------------------------------------------+-----------+--------+------------
 3.1.0   | Release    | 2021/11/15   | pgrouting-3.1.3 | Linux-5.11.0-40-generic | PostgreSQL 13.5 (Ubuntu 13.5-1.pgdg20.04+1) | GNU-9.3.0 | 1.71.0 | 3b5d8d3183
(1 row)
```

```
psql (13.7 (Ubuntu 13.7-1.pgdg20.04+1))
Type "help" for help.

sampledata=# SELECT * FROM pgr_bdDijkstra(
    'SELECT id, source, target, cost, reverse_cost FROM edge_table',
    2, 3
);
 seq | path_seq | node | edge | cost | agg_cost 
-----+----------+------+------+------+----------
   1 |        1 |    2 |    4 |    1 |        0
   2 |        2 |    5 |    8 |    1 |        1
   3 |        3 |    6 |    9 |    1 |        2
   4 |        4 |    9 |   16 |    1 |        3
   5 |        5 |    4 |    3 |    1 |        4
   6 |        6 |    3 |   -1 |    0 |        5
(6 rows)

sampledata=# SELECT * FROM pgr_bdDijkstra(
    'SELECT id, source, target, cost, reverse_cost FROM edge_table',
    2, 20
);
 seq | path_seq | node | edge | cost | agg_cost 
-----+----------+------+------+------+----------
(0 rows)
```

Looking deeper into the differences between versions I found this:
On version 3.1.2 there was a change on the bdDijkstra code:

```
git diff v3.1.2 v3.1.0 src/bdDijkstra/
diff --git a/src/bdDijkstra/bdDijkstra_driver.cpp b/src/bdDijkstra/bdDijkstra_driver.cpp
index 8707570bb9..1cd9a27261 100644
--- a/src/bdDijkstra/bdDijkstra_driver.cpp
+++ b/src/bdDijkstra/bdDijkstra_driver.cpp
@@ -71,17 +71,6 @@ pgr_bdDijkstra(
     std::deque<Path> paths;
     for (const auto source : sources) {
         for (const auto target : targets) {
-            if (source == target) {
-                paths.push_back(Path(source, target));
-                continue;
-            }
-
-            if (!graph.has_vertex(source)
-                    || !graph.has_vertex(target)) {
-                paths.push_back(Path(source, target));
-                continue;
-            }
-
             fn_bdDijkstra.clear();
             paths.push_back(fn_bdDijkstra.pgr_bdDijkstra(
                         graph.get_V(source), graph.get_V(target), only_cost));
```

From reading this code this issue has been fixed.
It was never opened per-se as a problem of bdDijkstra but more of a general problem in a variety of functions

It was fixed on PR #1770 that fixed #1725
With this commit: d6b32f16b80106fc626b157d174acd9ecdac5116

So the solution is, upgrade pgRouting to at least 3.1.2, (preferably the latest version)
